File: apps/mobile/android/app/src/main/java/com/aurelia/app/ui/PlayerViewModel.kt
package com.aurelia.app.ui

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.aurelia.app.auth.AuthInterceptor
import com.aurelia.app.data.model.Lyrics
import com.aurelia.app.data.model.LyricsAgent
import com.aurelia.app.data.model.LyricsSection
import com.aurelia.app.data.model.SyncedLine
import com.aurelia.app.data.model.SyncedWord
import com.aurelia.app.player.PlayerController
import com.aurelia.app.player.PlayerSnapshot
import com.aurelia.app.storage.SessionStore
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import uniffi.aurelia_core.getParsedLyrics as uniffiGetParsedLyrics
import uniffi.aurelia_core.markItemPlayed as uniffiMarkItemPlayed
import uniffi.aurelia_core.toggleFavorite as uniffiToggleFavorite
import uniffi.aurelia_lyrics.ParsedLyrics
import uniffi.aurelia_lyrics.ParsedLyricsLine

class PlayerViewModel(
    private val playerController: PlayerController,
    private val sessionStore: SessionStore,
) : ViewModel() {
    private val mutableState = MutableStateFlow(PlayerState())
    val state: StateFlow<PlayerState> = mutableState

    private var lastFetchedSongId: String? = null
    private var favoriteCache: MutableMap<String, Boolean> = mutableMapOf()

    init {
        playerController.observe { snapshot ->
            val previousSongId = mutableState.value.currentSongId
            val newSongId = snapshot.currentSongId

            // Update favorite cache from queue songs
            val queue = playerController.getQueue()
            queue.forEach { song ->
                if (!favoriteCache.containsKey(song.id)) {
                    favoriteCache[song.id] = song.isFavorite ?: false
                }
            }

            mutableState.update { current ->
                val isFavorite = newSongId?.let { favoriteCache[it] } ?: false
                current.fromSnapshot(snapshot, isFavorite)
            }

            // Mark the previous song as played when transitioning to a new song
            if (previousSongId != null && newSongId != null && newSongId != previousSongId) {
                markSongAsPlayed(previousSongId)
            }

            if (newSongId != null && newSongId.isNotBlank() && newSongId != lastFetchedSongId) {
                lastFetchedSongId = newSongId
                fetchLyrics(newSongId, snapshot.artist, snapshot.title)
            }
        }
    }

    private fun markSongAsPlayed(songId: String) {
        val serverUrl = sessionStore.getServerUrl() ?: return
        val token = sessionStore.getToken() ?: return
        val userId = sessionStore.getUserId() ?: return

        viewModelScope.launch(Dispatchers.IO) {
            try {
                uniffiMarkItemPlayed(serverUrl, token, userId, songId)
            } catch (e: Exception) {
                android.util.Log.e("PlayerViewModel", "Failed to mark song as played: $songId", e)
            }
        }
    }

    private fun fetchLyrics(
        songId: String?,
        artist: String,
        title: String,
    ) {
        viewModelScope.launch(Dispatchers.IO) {
            lastFetchedSongId = songId
            mutableState.update { it.copy(lyrics = null) }
            try {
                val serverUrl = sessionStore.getServerUrl() ?: ""
                val token = sessionStore.getToken() ?: ""
                val itemId = songId ?: ""

                val lyricsServerUrl = sessionStore.getLyricsServerUrl()
                val lyrics =
                    uniffiGetParsedLyrics(serverUrl, token, itemId, artist, title, null, lyricsServerUrl).toUiLyrics()

                // Ensure we are still playing the same song
                if (songId == mutableState.value.currentSongId) {
                    if (lyrics.isValid()) {
                        mutableState.update { it.copy(lyrics = lyrics) }
                    } else {
                        mutableState.update { it.copy(showLyrics = false) }
                    }
                }
            } catch (e: Exception) {
                if (songId == mutableState.value.currentSongId) {
                    mutableState.update { it.copy(showLyrics = false) }
                }
                // Only handle auth errors, ignore others (e.g., lyrics not found)
                AuthInterceptor.handlePotentialAuthError(e)
            }
        }
    }

    fun togglePlayPause() {
        val currentState = mutableState.value
        if (currentState.isPlaying) {
            playerController.pause()
        } else {
            playerController.resume()
        }
    }

    fun seekTo(positionMs: Long) {
        playerController.seekTo(positionMs)
    }

    fun skipNext() {
        playerController.skipNext()
    }

    fun skipPrevious() {
        playerController.skipPrevious()
    }

    private fun PlayerState.fromSnapshot(
        snapshot: PlayerSnapshot,
        isFavorite: Boolean,
    ): PlayerState =
        copy(
            title = snapshot.title,
            artist = snapshot.artist,
            albumArtUrl = snapshot.albumArtUrl,
            isPlaying = snapshot.isPlaying,
            isBuffering = snapshot.isBuffering,
            positionMs = snapshot.positionMs,
            durationMs = snapshot.durationMs,
            queue = playerController.getQueue(),
            currentQueueIndex = playerController.getCurrentQueueIndex(),
            isShuffled = snapshot.isShuffled,
            repeatMode = snapshot.repeatMode,
            currentSongId = snapshot.currentSongId,
            currentAlbumId = snapshot.currentAlbumId,
            currentArtistId = snapshot.currentArtistId,
            currentAlbumName = snapshot.currentAlbumName,
            isFavorite = isFavorite,
            playbackSpeed = snapshot.playbackSpeed,
            updateTimeMs = snapshot.updateTimeMs,
            codec = snapshot.codec,
            bitRate = snapshot.bitRate,
            sampleRate = snapshot.sampleRate,
        )

    private fun ParsedLyricsLine.toUiSyncedLine(): SyncedLine =
        SyncedLine(
            time = timeMs.toInt(),
            endTime = endTimeMs?.toInt(),
            line = line,
            words = words?.map { word ->
                SyncedWord(
                    time = word.timeMs.toInt(),
                    endTime = word.endTimeMs?.toInt(),
                    word = word.word,
                )
            },
            agentId = agentId,
            translation = translation,
        )

    private fun ParsedLyrics.toUiLyrics(): Lyrics {
        val syncedLines =
            synced
                .takeIf { it.isNotEmpty() }
                ?.map { it.toUiSyncedLine() }

        val uiSections = sections?.takeIf { it.isNotEmpty() }?.map { section ->
            LyricsSection(
                name = section.name,
                startTime = section.startTimeMs.toInt(),
                endTime = section.endTimeMs.toInt(),
                lines = section.lines.map { it.toUiSyncedLine() },
                agentId = section.agentId,
            )
        }

        val uiAgents = agents?.takeIf { it.isNotEmpty() }?.map { agent ->
            LyricsAgent(id = agent.id, agentType = agent.agentType)
        }

        val plainLines = plain.takeIf { it.isNotEmpty() }

        return Lyrics(
            plain = plainLines,
            synced = syncedLines,
            sections = uiSections,
            agents = uiAgents,
            songwriters = songwriters?.takeIf { it.isNotEmpty() },
            language = language,
            areFromRemote = areFromRemote,
        )
    }

    fun playQueueItem(index: Int) {
        playerController.playQueueItem(index)
    }

    fun toggleLyrics() {
        val currentState = mutableState.value
        if (currentState.lyrics == null && currentState.title.isNotBlank()) {
            fetchLyrics(currentState.currentSongId, currentState.artist, currentState.title)
        }
        mutableState.update { it.copy(showLyrics = !it.showLyrics) }
    }

    fun setLyrics(lyrics: Lyrics?) {
        mutableState.update { it.copy(lyrics = lyrics) }
    }

    fun clearLyrics() {
        mutableState.update { it.copy(lyrics = null, showLyrics = false) }
    }

    fun toggleShuffle() {
        playerController.toggleShuffle()
    }

    fun cycleRepeatMode() {
        playerController.cycleRepeatMode()
    }

    fun toggleFavorite() {
        val currentState = mutableState.value
        val songId = currentState.currentSongId ?: return

        val serverUrl = sessionStore.getServerUrl() ?: return
        val token = sessionStore.getToken() ?: return
        val userId = sessionStore.getUserId() ?: return
        val targetFavoriteState = !currentState.isFavorite

        mutableState.update { it.copy(isFavoriteLoading = true) }

        viewModelScope.launch(Dispatchers.IO) {
            try {
                val newFavoriteState =
                    uniffiToggleFavorite(
                        serverUrl = serverUrl,
                        token = token,
                        userId = userId,
                        itemId = songId,
                        isFavorite = targetFavoriteState,
                    )
                favoriteCache[songId] = newFavoriteState
                mutableState.update {
                    it.copy(isFavorite = newFavoriteState, isFavoriteLoading = false)
                }
            } catch (e: Exception) {
                if (!AuthInterceptor.handlePotentialAuthError(e)) {
                    android.util.Log.e("PlayerViewModel", "Failed to toggle favorite", e)
                    mutableState.update { it.copy(isFavoriteLoading = false) }
                }
            }
        }
    }

    fun setFavoriteFromSong(
        songId: String,
        isFavorite: Boolean,
    ) {
        favoriteCache[songId] = isFavorite
        if (mutableState.value.currentSongId == songId) {
            mutableState.update { it.copy(isFavorite = isFavorite) }
        }
    }
}

File: apps/mobile/ios/Aurelia/ViewModels/PlayerViewModel.swift
import Foundation
import Observation
import os
import AureliaCore

@Observable
final class PlayerViewModel: @unchecked Sendable {
    var title = ""
    var artist = ""
    var albumArtUrl: String?
    var isPlaying = false
    var isBuffering = false
    var positionMs: Int64 = 0
    var durationMs: Int64 = 0
    var positionUpdateTimeMs: Int64 = 0
    var queue: [Song] = []
    var currentQueueIndex: Int = -1
    var lyrics: Lyrics?
    var showLyrics = false
    var isShuffled = false
    var repeatMode: RepeatMode = .none
    var currentSongId: String?
    var currentAlbumId: String?
    var currentArtistId: String?
    var currentAlbumName: String?
    var isFavorite = false
    var isFavoriteLoading = false
    var playbackSpeed: Float = 1.0
    var codec: String?
    var bitRate: Int32?
    var sampleRate: Int32?

    var hasPrevious: Bool { currentQueueIndex > 0 }
    var hasNext: Bool { currentQueueIndex >= 0 && currentQueueIndex < queue.count - 1 }

    var formatInfo: String? {
        var parts: [String] = []
        if let codec { parts.append(codec.uppercased()) }
        if let sampleRate { parts.append("\(Double(sampleRate) / 1000.0) kHz") }
        if let bitRate { parts.append("\(bitRate / 1000) kbps") }
        return parts.isEmpty ? nil : parts.joined(separator: " / ")
    }

    private let sessionStore = SessionStore.shared
    private let logger = Logger(subsystem: "com.aurelia.app", category: "PlayerViewModel")
    private var lastFetchedSongId: String?
    private var favoriteCache: [String: Bool] = [:]

    /// Call this to sync state from the player controller.
    func updateFrom(_ snapshot: PlayerSnapshot, position: PlaybackPosition, playerController: AudioPlayerController) {
        let previousSongId = currentSongId
        let newSongId = snapshot.currentSongId
        let didChangeSong = newSongId != previousSongId
        let shouldRefreshQueue = queue.isEmpty || didChangeSong || snapshot.isShuffled != isShuffled

        if shouldRefreshQueue {
            let latestQueue = playerController.getQueue()
            if queueSignature(for: latestQueue) != queueSignature(for: queue) {
                queue = latestQueue
            }

            // Update favorite cache from queue only when queue changes.
            for song in latestQueue where favoriteCache[song.id] == nil {
                favoriteCache[song.id] = song.isFavorite ?? false
            }
        }

        let latestQueueIndex = playerController.getCurrentQueueIndex()

        setIfChanged(\.title, snapshot.title)
        setIfChanged(\.artist, snapshot.artist)
        setIfChanged(\.albumArtUrl, snapshot.albumArtUrl)
        setIfChanged(\.isPlaying, snapshot.isPlaying)
        setIfChanged(\.isBuffering, snapshot.isBuffering)
        setIfChanged(\.positionMs, position.positionMs)
        setIfChanged(\.durationMs, position.durationMs)
        setIfChanged(\.positionUpdateTimeMs, position.updateTimeMs)
        setIfChanged(\.currentQueueIndex, latestQueueIndex)
        setIfChanged(\.isShuffled, snapshot.isShuffled)
        setIfChanged(\.repeatMode, snapshot.repeatMode)
        setIfChanged(\.currentSongId, newSongId)
        setIfChanged(\.currentAlbumId, snapshot.currentAlbumId)
        setIfChanged(\.currentArtistId, snapshot.currentArtistId)
        setIfChanged(\.currentAlbumName, snapshot.currentAlbumName)
        setIfChanged(\.isFavorite, newSongId.flatMap { favoriteCache[$0] } ?? false)
        setIfChanged(\.playbackSpeed, snapshot.playbackSpeed)
        setIfChanged(\.codec, snapshot.codec)
        setIfChanged(\.bitRate, snapshot.bitRate)
        setIfChanged(\.sampleRate, snapshot.sampleRate)

        // Mark previous song as played on track change
        if let previousSongId, didChangeSong, newSongId != nil {
            markSongAsPlayed(previousSongId)
        }

        // Fetch lyrics for new song
        if let newSongId, !newSongId.isEmpty, didChangeSong, newSongId != lastFetchedSongId {
            lastFetchedSongId = newSongId
            fetchLyrics(songId: newSongId, artist: snapshot.artist, title: snapshot.title)
        }
    }

    // MARK: - Playback Controls

    func togglePlayPause(playerController: AudioPlayerController) {
        if isPlaying {
            playerController.pause()
        } else {
            playerController.resume()
        }
    }

    func seekTo(_ positionMs: Int64, playerController: AudioPlayerController) {
        playerController.seekTo(positionMs: positionMs)
    }

    func skipNext(playerController: AudioPlayerController) {
        playerController.skipNext()
    }

    func skipPrevious(playerController: AudioPlayerController) {
        playerController.skipPrevious()
    }

    func playQueueItem(_ index: Int, playerController: AudioPlayerController) {
        playerController.playQueueItem(index)
    }

    func toggleShuffle(playerController: AudioPlayerController) {
        playerController.toggleShuffle()
    }

    func cycleRepeatMode(playerController: AudioPlayerController) {
        playerController.cycleRepeatMode()
    }

    // MARK: - Lyrics

    func toggleLyrics() {
        if lyrics == nil && !title.isEmpty {
            fetchLyrics(songId: currentSongId, artist: artist, title: title)
        }
        showLyrics.toggle()
    }

    private func fetchLyrics(songId: String?, artist: String, title: String) {
        lyrics = nil
        
        Task { @MainActor [self] in
            let serverUrl = self.sessionStore.serverUrl ?? ""
            let token = self.sessionStore.token ?? ""
            let lyricsServerUrl = self.sessionStore.lyricsServerUrl
            let itemId = songId ?? ""

            self.logger.info("[Lyrics] Fetching lyrics for '\(title)' by '\(artist)' (itemId=\(itemId), serverUrl=\(serverUrl.prefix(30))..., hasToken=\(!token.isEmpty))")

            self.logger.info("[Lyrics] Debug: lyricsServerUrl = '\(lyricsServerUrl ?? "nil")'")
            let parsedLyrics = await getParsedLyrics(
                serverUrl: serverUrl,
                token: token,
                itemId: itemId,
                artist: artist,
                title: title,
                path: nil,
                lyricsServerUrl: lyricsServerUrl
            )

            self.logger.info("[Lyrics] Got ParsedLyrics from core: syncedLines=\(parsedLyrics.synced.count), plainLines=\(parsedLyrics.plain.count), areFromRemote=\(parsedLyrics.areFromRemote), hasSections=\(parsedLyrics.sections != nil), hasAgents=\(parsedLyrics.agents != nil), hasSongwriters=\(parsedLyrics.songwriters != nil), language=\(parsedLyrics.language ?? "nil")")

            // Log first few synced lines for debugging
            for (i, line) in parsedLyrics.synced.prefix(5).enumerated() {
                let wordCount = line.words?.count ?? 0
                self.logger.info("[Lyrics] synced[\(i)]: timeMs=\(line.timeMs), endTimeMs=\(line.endTimeMs.map { String($0) } ?? "nil"), words=\(wordCount), agentId=\(line.agentId ?? "nil"), text='\(line.line.prefix(60))'")
                if let words = line.words {
                    for (j, word) in words.prefix(3).enumerated() {
                        self.logger.info("[Lyrics] word[\(j)]: timeMs=\(word.timeMs), endTimeMs=\(word.endTimeMs.map { String($0) } ?? "nil"), '\(word.word)'")
                    }
                    if words.count > 3 {
                        self.logger.info("[Lyrics] ... and \(words.count - 3) more words")
                    }
                }
            }
            if parsedLyrics.synced.count > 5 {
                self.logger.info("[Lyrics] ... and \(parsedLyrics.synced.count - 5) more synced lines")
            }

            let parsed = LyricsParser.fromParsed(parsedLyrics)
            let isValid = parsed.isValid
            let hasSynced = parsed.synced != nil
            let syncedCount = parsed.synced?.count ?? 0
            let hasPlain = parsed.plain != nil
            let hasWordSync = parsed.synced?.first?.words != nil
            self.logger.info("[Lyrics] After LyricsParser: isValid=\(isValid), hasSynced=\(hasSynced), syncedCount=\(syncedCount), hasPlain=\(hasPlain), hasWordSync=\(hasWordSync)")

            if songId == self.currentSongId {
                if isValid {
                    self.logger.info("[Lyrics] Setting lyrics on view model (valid)")
                    self.lyrics = parsed
                } else {
                    self.logger.warning("[Lyrics] Lyrics invalid, hiding lyrics panel")
                    self.showLyrics = false
                }
        } else {
            self.logger.info("[Lyrics] Song changed during fetch (was=\(songId ?? "nil"), now=\(self.currentSongId ?? "nil")), discarding result")
        }
    }
}

// MARK: - Favorites

    func toggleFavorite() {
        guard let songId = currentSongId else { return }
        guard let serverUrl = sessionStore.serverUrl,
              let token = sessionStore.token,
              let userId = sessionStore.userId else { return }
        let targetFavoriteState = !isFavorite

        isFavoriteLoading = true

        Task.detached { [self, targetFavoriteState] in
            do {
                let newState = try await AureliaCore.toggleFavorite(
                    serverUrl: serverUrl,
                    token: token,
                    userId: userId,
                    itemId: songId,
                    isFavorite: targetFavoriteState
                )
                await MainActor.run {
                    self.favoriteCache[songId] = newState
                    self.isFavorite = newState
                    self.isFavoriteLoading = false
                }
            } catch {
                if await !AuthInterceptor.shared.handlePotentialAuthError(error) {
                    self.logger.error("Failed to toggle favorite: \(error)")
                    await MainActor.run { self.isFavoriteLoading = false }
                }
            }
        }
    }

    // MARK: - Scrobble

    private func setIfChanged<T: Equatable>(_ keyPath: ReferenceWritableKeyPath<PlayerViewModel, T>, _ value: T) {
        if self[keyPath: keyPath] != value {
            self[keyPath: keyPath] = value
        }
    }

    private func queueSignature(for queue: [Song]) -> String {
        guard let first = queue.first?.id, let last = queue.last?.id else {
            return "empty:\(queue.count)"
        }
        return "\(queue.count):\(first):\(last)"
    }

    private func markSongAsPlayed(_ songId: String) {
        guard let serverUrl = sessionStore.serverUrl,
              let token = sessionStore.token,
              let userId = sessionStore.userId else { return }

        Task.detached {
            do {
                try await markItemPlayed(
                    serverUrl: serverUrl,
                    token: token,
                    userId: userId,
                    itemId: songId
                )
            } catch {
                self.logger.error("Failed to mark song as played: \(songId) - \(error)")
            }
        }
    }
}
